File: frontend/src/utils/planTarget.ts
export function normalizePlanTarget(hours: unknown, minutes: unknown) {
  const parsedHours = Number(hours)
  const parsedMinutes = Number(minutes)

  return {
    fasting_hours: Number.isFinite(parsedHours)
      ? Math.max(0, Math.floor(parsedHours))
      : 0,
    fasting_minutes: Number.isFinite(parsedMinutes)
      ? Math.min(59, Math.max(0, Math.floor(parsedMinutes)))
      : 0,
  }
}

export function totalTargetMinutes(hours: number, minutes: number): number {
  return hours * 60 + minutes
}

export function isValidPlanTarget(hours: number, minutes: number): boolean {
  return (
    hours >= 0 &&
    minutes >= 0 &&
    minutes <= 59 &&
    totalTargetMinutes(hours, minutes) >= 1
  )
}

export function formatTargetDuration(hours: number, minutes: number): string {
  if (hours === 0) return `${minutes}dk`
  if (minutes === 0) return `${hours}s`
  return `${hours}s ${minutes}dk`
}

export function completionFromDuration(
  durationHours: number,
  targetHours: number,
  targetMinutes: number,
) {
  const targetTotal = totalTargetMinutes(targetHours, targetMinutes)
  const elapsedMinutes = durationHours * 60
  const percent =
    targetTotal > 0 ? Math.round((elapsedMinutes / targetTotal) * 100) : 0

  return {
    percent,
    fillRatio: targetTotal > 0 ? Math.min(elapsedMinutes / targetTotal, 1) : 0,
  }
}

File: ios/Pacor/Utils/PlanTarget.swift
import Foundation

enum PlanTarget {
    static func normalize(hours: Int, minutes: Int) -> (fastingHours: Int, fastingMinutes: Int) {
        (
            fastingHours: max(0, hours),
            fastingMinutes: min(59, max(0, minutes))
        )
    }

    static func totalTargetMinutes(hours: Int, minutes: Int) -> Int {
        hours * 60 + minutes
    }

    static func isValid(hours: Int, minutes: Int) -> Bool {
        hours >= 0 && minutes >= 0 && minutes <= 59 && totalTargetMinutes(hours: hours, minutes: minutes) >= 1
    }

    static func formatTargetDuration(hours: Int, minutes: Int) -> String {
        if hours == 0 { return "\(minutes)dk" }
        if minutes == 0 { return "\(hours)s" }
        return "\(hours)s \(minutes)dk"
    }

    static func completion(fromDurationHours durationHours: Double, targetHours: Int, targetMinutes: Int) -> CompletionInfo {
        let targetTotal = totalTargetMinutes(hours: targetHours, minutes: targetMinutes)
        let elapsedMinutes = durationHours * 60
        let percent = targetTotal > 0 ? Int(round((elapsedMinutes / Double(targetTotal)) * 100)) : 0
        let fillRatio = targetTotal > 0 ? min(elapsedMinutes / Double(targetTotal), 1) : 0
        return CompletionInfo(percent: percent, fillRatio: fillRatio)
    }

    static func prepareForSave(_ plan: PlanSettings) -> PlanSettings {
        let normalized = normalize(hours: plan.fastingHours, minutes: plan.fastingMinutes)
        let eatingHours = max(1, 24 - normalized.fastingHours)
        return PlanSettings(
            planType: "custom",
            eatingHours: eatingHours,
            fastingHours: normalized.fastingHours,
            fastingMinutes: normalized.fastingMinutes
        )
    }
}
